### PR TITLE
Support opaque Rust types in a rust::Slice

### LINKED ISSUE #532: Support opaque Rust types in a rust::Slice
Currently only primitives and shared types are supported in a slice in the FFI.

The following is currently rejected, but it should be possible to support:

```rust
#[cxx::bridge]
mod ffi {
    extern "Rust" {
        type MyType;
    }

    unsafe extern "C++" {
        fn f(slice: &[MyType]);
    }
}
```

The slice iterators and indexing operations will need to obtain a `stride` at runtime from Rust, exactly like we already do for rust::Vec containing opaque Rust types.

### PR DESCRIPTION
**Pull request by MichaelRFairhurst**
Fixes #532.

### CODE REVIEW

**Review by dtolnay — CHANGES_REQUESTED** [review]
Thanks, this looks nice.

One thing I am concerned about with this implementation is that the coupling between slice and vec impls will cause poor interaction with #496. For example if we have a hierarchy of two crates where upstream only involves &amp;\[T\] while downstream wants to use Vec\<T\>, and they write:

```rust
// dependency

#[cxx::bridge]
mod ffi {
    extern "Rust" {
        type Thing;

        fn f(slice: &[Thing]);
    }
}
```

```rust
// downstream

#[cxx::bridge]
mod ffi {
    use dependency::Thing;

    extern "Rust" {
        fn g() -> Vec<Thing>;
    }

    impl Vec<Thing> {}
}
```

then they end up with a linker error from duplicating the stride symbol, and "please synthesize a Vec shim *except* the stride symbol" is a confusing thing for the bridge author to need to express.

Decoupling vec and slice such that a functionally identical stride symbol is emitted for each under a different name would fix that, but I wonder if a better fix might be to separate stride from vec/slice altogether and emit a stride accessor unconditionally for every opaque Rust type. Having a size available is going to be generally useful for lots of things on the C++ side as we grow the C++ API of the bindings in cxx.h. It should be fine for C++ to be able to know the size of every type, whether used in vec/slice or not, and whether at compile time (for shared structs, via `sizeof`) or only runtime (for opaque Rust types).

What I'd like to support on the C++ side looks like:

```rust
// include/example.h

#pragma once
#include "rust/cxx.h"

struct Thing;

void f(rust::Slice<Thing> slice);  // forward decl is sufficient i.e. does not
                                   // require knowing the size yet at this point
```

```rust
// src/example.cc

#include "include/example.h"
#include "src/main.rs.h"  // the bridge generated C++ code

void f(rust::Slice<Thing> slice) {
  (void) slice.at(1);  // requires knowing the size of Thing as described below
}
```

where `at` (and similarly `operator[]`, `begin` etc) has a SFINAE to use sizeof(T) if sizeof(T) is not a substitution failure i.e. a definition of T is available and not just a forward decl like for opaque types, otherwise makes a runtime call to find out the size from Rust, otherwise fails to compile if that call is not available.

So for input like this:

```rust
#[cxx::bridge]
mod ffi {
    extern "Rust" {
        type Thing;
    }
}
```

the code generator produces:

```cpp
struct Thing;
namespace detail {
  // every opaque Rust type gets an overload of this:
  inline size_t stride(Thing *) { return cxxbridge1$stride$Thing(); }
}
```

and `at` does something like (untested):

```cpp
namespace detail {
template <typename T, typename = size_t>
struct rust_sizeof {
  size_t operator()() { return stride(static_cast<T *>(nullptr)); }
};
template <typename T>
struct rust_sizeof<T, decltype(sizeof(T))> {  // partial template specialization
  size_t operator()() { return sizeof(T); }
};
} // namespace detail

template <typename T>
T &Slice<T>::operator[](size_t n) noexcept {
  size_t stride = detail::rust_sizeof<T>{}();
  ...
}
```

File: include/cxx.h [hunk 1]
```diff
@@ -502,16 +506,31 @@ T *Slice<T>::iterator::operator->() const noexcept {
...
+    panic<std::out_of_range>("rust::Slice index out of range");
+  }
+  return (*this)[n];
+}
+
+
 template <typename T>
 typename Slice<T>::iterator &Slice<T>::iterator::operator++() noexcept {
-  ++this->pos;
+  this->pos += this->stride;
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply include/cxx.h:526]: Fixed

File: macro/src/expand.rs [hunk 2]
```diff
@@ -112,6 +112,27 @@ fn expand(ffi: Module, apis: &[Api], types: &Types) -> TokenStream {
...
+                Some(ident)
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }).collect::<std::collections::HashSet<_>>();
+    for ty in strided_types {
+        hidden.extend(expand_rust_stride(ty, types));
+    }
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply macro/src/expand.rs:134]: Good point! With the new approach this is no longer used.

File: macro/src/expand.rs [hunk 3]
```diff
@@ -112,6 +112,27 @@ fn expand(ffi: Module, apis: &[Api], types: &Types) -> TokenStream {
...
     }
 
+    // Generate strides
+    let strided_types = types.into_iter().filter_map(|ty| match ty {
+        Type::RustVec(ty) => Some(&ty.inner),
+        Type::SliceRef(ty) => Some(&ty.inner),
+        _ => None,
+    }).filter_map(|ty| match ty {
+        Type::Ident(ident) => {
+            let explicit_impl = types.explicit_impls.get(ty);
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply macro/src/expand.rs:122]: Good catch, subtle and important.

With the changes as you suggested, this is fixed.

**Comment by MichaelRFairhurst** [comment]
New version up for review.

I ran into a few things that make SFINAE become slightly more complicated; nothing super major but we did talk about some over chat, and I ended up feeling like a simple sizeof<T>() template function was the right first approach.

Essentially, the SFINAE approach requires a forward declaration of

```rust
template <typename T, typename = size_t>
struct rust_sizeof; // forward declaration
template <typename T>
struct rust_sizeof<T, decltype(sizeof(T))> {  // partial template specialization
  size_t operator()() { return sizeof(T); }
};
```

because the `stride()` functions we wish to call have not been generated yet. To then be able to call `operator[]` in a `.cpp` file, the `.rs.h` file must define `rust_sizeof` also at least forward declare `Stride(T*)` for all types, independent from declaring their definitions and extern functions. This does not fit with other patterns where generic instantiations are only generated in `.rs.cpp`. These problems are _minor_, I just got a growing impression that the code I was writing did not quite fit the current project as well, and stashed that in a branch before getting the PR to the current state. Happy to revisit this, perhaps after landing this or perhaps before.

One thing I noticed about the PR in this state which may be an issue is that this code:

```
#[cxx::bridge]
mod ffi {
    extern "Rust" {
        type Thing;
    }
}
```

Does not produce a `sizeof` symbol for `Thing`, because it is not a part of `Types.all` (though it is a part of `Types.rust` -- which is a different set type). It will however get a `sizeof` symbol if it is referenced from any APIs (thereby becoming a part of `Types.all`), so I don't think this is a critical path of urgent support or reasonable to rely on in terms of nonlocal types.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: gen/src/write.rs [hunk 4]
```diff
@@ -1366,6 +1382,36 @@ fn write_rust_box_extern(out: &mut OutFile, ident: &Pair) {
...
+
+fn write_rust_sizeof_impl(out: &mut OutFile, element: &RustName) {
+    let instance = element.to_mangled(out.types);
+    let typename = element.to_typename(out.types);
+
+    writeln!(out, "#ifndef CXX_BRIDGE_RUST_SIZEOF_IMPL_{}", instance);
+    writeln!(out, "#define CXX_BRIDGE_RUST_SIZEOF_IMPL_{}", instance);
+    out.begin_block(Block::Namespace("detail"));
+    writeln!(out, "template<>");
+    writeln!(out, "::std::size_t rust_sizeof<{}>() {{", typename);
```
Comment (dtolnay) [inline gen/src/write.rs:1403]: I am hesitant about over-reliance on the linker for resolving these template specializations. If the user winds up instantiating `rust::Slice` in C++ with some type that they shouldn't, in this implementation they would only find that out as a link error, not a compile error, due to the template declaration (without definition) `template <typename T> size_t rust_sizeof()` present in include/cxx.h. In general deferring errors from compile time to link time makes them harder to debug.

I found a different way to do this which doesn't use undefined template declarations and is able to surface all errors at compile time: see #595 + #597.

File: include/cxx.h [hunk 5]
```diff
@@ -156,6 +164,8 @@ class Slice final
   T *data() const noexcept;
   std::size_t size() const noexcept;
   std::size_t length() const noexcept;
+  T &operator[](std::size_t n) noexcept;
+  T &at(std::size_t n);
```
Comment (dtolnay) [inline include/cxx.h:168]: These changes are not related to opaque type support so I've pulled them out to #598.

File: include/cxx.h [hunk 6]
```diff
@@ -192,6 +203,7 @@ class Slice<T>::iterator final {
 private:
   friend class Slice;
   T *pos;
+  std::size_t stride;
```
Comment (dtolnay) [inline include/cxx.h:206]: Extracted to #600 + #601.

File: syntax/check.rs [hunk 7]
```diff
@@ -210,16 +210,15 @@ fn check_type_ref(cx: &mut Check, ty: &Ref) {
 fn check_type_slice_ref(cx: &mut Check, ty: &SliceRef) {
     let supported = match &ty.inner {
         Type::Str(_) | Type::SliceRef(_) => false,
+        Type::Ident(ident) => !is_opaque_cxx(cx, &ident.rust),
```
Comment (dtolnay) [inline syntax/check.rs:213]: Extracted to #603.

File: src/symbols/rust_vec.rs [hunk 8]
```diff
@@ -53,12 +53,6 @@ macro_rules! rust_vec_shims {
                     (*this).repr.set_len(len);
                 }
             }
-            attr! {
-                #[export_name = concat!("cxxbridge1$rust_vec$", $segment, "$stride")]
-                unsafe extern "C" fn __stride() -> usize {
-                    mem::size_of::<$ty>()
-                }
-            }
```
Comment (dtolnay) [inline src/symbols/rust_vec.rs:61]: Extracted to #604.

File: src/symbols/rust_sizeof.rs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+rust_stride_shims_for_primitive!(u64);
+rust_stride_shims_for_primitive!(i8);
+rust_stride_shims_for_primitive!(i16);
+rust_stride_shims_for_primitive!(i32);
+rust_stride_shims_for_primitive!(i64);
+rust_stride_shims_for_primitive!(f32);
+rust_stride_shims_for_primitive!(f64);
+
+rust_stride_shims!("char", c_char);
+rust_stride_shims!("string", RustString);
```
Comment (dtolnay) [inline src/symbols/rust_sizeof.rs:41]: The alternative approach in #597 makes these not needed.

**Review by dtolnay — COMMENTED** [review]
I merged this with a parallel set of fixes linked below.

Thanks!

#594, #595, #596, #597, #598, #599, #600, #601, #602, #603, #604, #605

File: include/cxx.h [hunk 10]
```diff
@@ -551,13 +584,15 @@ template <typename T>
 typename Slice<T>::iterator Slice<T>::begin() const noexcept {
   iterator it;
   it.pos = this->ptr;
+  it.stride = this->stride();
```
[the remark this answers is not in the document]
Comment (dtolnay) [reply include/cxx.h:587]: Fixed in #602.

File: tests/ffi/lib.rs [hunk 11]
```diff
@@ -120,6 +120,8 @@ pub mod ffi {
         fn c_take_str(s: &str);
         fn c_take_slice_char(s: &[c_char]);
         fn c_take_slice_shared(s: &[Shared]);
+        fn c_take_slice_not_usized(s: &[Second]);
+        fn c_take_slice_not_usized_mut(s: &mut [Second]);
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply tests/ffi/lib.rs:124]: I meant sizeof(T) != sizeof(usize), I can see how that is unclear! :)

File: tests/ffi/tests.cc [hunk 12]
```diff
@@ -274,7 +274,19 @@ void c_take_slice_char(rust::Slice<const char> s) {
...
+void c_take_slice_not_usized(rust::Slice<const second::Second> s) {
+  if (s.size() == 2 && s[0].i == 1 && s[0].e == tests::COwnedEnum::CVAL1 && s[1].i == 2 && s[1].e == COwnedEnum::CVAL2) {
+    cxx_test_suite_set_correct();
+  }
+}
+
+void c_take_slice_not_usized_mut(rust::Slice<second::Second> s) {
+  if (s.size() == 2 && s[0].i == 1 && s[0].e == tests::COwnedEnum::CVAL1 && s[1].i == 2 && s[1].e == COwnedEnum::CVAL2) {
     cxx_test_suite_set_correct();
   }
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply tests/ffi/tests.cc:291]: Correct -- I could and should have updated this accordingly!

File: macro/src/expand.rs [hunk 13]
```diff
@@ -109,6 +109,12 @@ fn expand(ffi: Module, apis: &[Api], types: &Types) -> TokenStream {
                     expanded.extend(expand_cxx_vector(ident, explicit_impl, types));
                 }
             }
+        } else if let Type::Ident(ident) = ty {
+            let explicit_impl = types.explicit_impls.get(ty);
+            if Atom::from(&ident.rust).is_none()
+                    && (explicit_impl.is_some() || !types.aliases.contains_key(&ident.rust)) {
+                hidden.extend(expand_rust_sizeof(ident, types));
+            }
```
[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply macro/src/expand.rs:117]: I saw you chose instead to emit for "RustType()" in "apis." I was wondering if that was the best approach, good to get feedback that it is! :)

File: tests/ffi/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (dtolnay) [reply tests/ffi/lib.rs:124]: Well sizeof(T) is actually the same as sizeof(usize) here.

```rust
println!("{}", std::mem::size_of::<Second>());
println!("{}", std::mem::size_of::<usize>());
```

```console
8
8
```

[the remark this answers is not in the document]
Comment (MichaelRFairhurst) [reply tests/ffi/lib.rs:124]: Good point!

**Pull request merged**